### PR TITLE
 [proxy] enable metrics for Caddy

### PR DESCRIPTION
**Pull request by kylos101**
## Description

Presently, we're only getting `caddy_reverse_proxy_upstreams_healthy` metric from caddy, which we cannot inspect to alert based on status code (like for 50x errors).

This change makes it possible to inspect caddy metrics like `caddy_http_response_duration_seconds_count{code=~"401|502"}`. 

## Related Issue(s)

Related to ENT-268

## How to test

From the preview, start a workspace, port-forward Grafana, and do `increase(caddy_http_response_duration_seconds_count{}[5m])`

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/proxy/conf/Caddyfile [hunk 1]
```diff
@@ -22,9 +22,18 @@
 
 	servers {
 		protocols h1 h2 h2c
+		metrics
```
Comment (kylos101) [inline components/proxy/conf/Caddyfile:25]: This adds various caddy_* metrics

File: components/proxy/conf/Caddyfile [hunk 2]
```diff
@@ -22,9 +22,18 @@
...
 	}
 }
 
+# Define a matcher to exclude 502 errors on a specific path
+@exclude502 {
```
Comment (kylos101) [inline components/proxy/conf/Caddyfile:30]: This'll exclude http requests from caddy metrics if the path and status match (I found while testing in preview that 502's were emitted when workspaces started successfully).

Metrics did not have a label with path, I had to look in logs, and there I saw they were all for supervisor, hence this rule being exclusive to supervisor.

@iQQBot what is the 502 for? I assume it is proxy inspecting supervisor endpoints for workspaces. How come `proxy` is checking the supervisor endpoint for workspaces? I was not expecting this and am just curious.

Comment (kylos101) [reply components/proxy/conf/Caddyfile:30]: Here are the 502's I am referring:
image

Comment (geropl) [reply components/proxy/conf/Caddyfile:30]: The 502 are forwarded by `ws-proxy` as long as the workspace is not up, yet, but we already try to poll the status (until it's there).

But note that to my knowledge Previews are the only setup where we still use the `proxy` -> `ws-proxy` chain (FYI originally we used it in self-hosted to just require a single IP).

Let's go with this for now, and see if we can refine it later.

**Review by geropl — APPROVED** [review]
Nice! :heavy_check_mark:

File: components/proxy/conf/Caddyfile (hunk 2 above)
Comment (kylos101) [reply components/proxy/conf/Caddyfile:30]: > But note that to my knowledge Previews are the only setup where we still use the proxy -> ws-proxy chain

@geropl Can you share code references to what you're referring? This seems wrong to me, we should try to have parity, if possible.

Comment (kylos101) [reply components/proxy/conf/Caddyfile:30]: Moving conversation to ENT-268

**Pull request merged**